sdk/java/runtime/main.go

```diff
@@ -0,0 +1,312 @@
+// Runtime module for the Java SDK
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"java-sdk/internal/dagger"
+
+	"github.com/iancoleman/strcase"
+)
+
+const (
+	MavenImage  = "maven:3.9.9-eclipse-temurin-23-alpine"
+	MavenDigest = "sha256:77fe6f79f868484d85679bfaa121b34177e3fedf570086ea6babbd6db2223e89"
+	JavaImage   = "eclipse-temurin:23-jre-alpine"
+	JavaDigest  = "sha256:bd8e2c8c19bcadbaa8c6a128051a22384c6f7cfe5fa520cb663fe21fff96f084"
+
+	ModSourceDirPath = "/src"
+	ModDirPath       = "/opt/module"
+	GenPath          = "sdk"
+)
+
+type JavaSdk struct {
+	SDKSourceDir  *dagger.Directory
+	RequiredPaths []string
+	moduleConfig  moduleConfig
+}
+
+type moduleConfig struct {
+	name    string
+	subPath string
+}
+
+func (c *moduleConfig) modulePath() string {
+	return filepath.Join(ModSourceDirPath, c.subPath)
+}
+
+func (c *moduleConfig) sdkPath() string {
+	return filepath.Join(c.modulePath(), GenPath)
+}
+
+func New(
+	// Directory with the Java SDK source code.
+	// dagger-java-samples is not necessary to build, but as it's referenced in the root pom.xml maven
+	// will check if it's there. So we keep it, until a better solution (like to extract the samples somewhere else)
+	// +optional
+	// +defaultPath="/sdk/java"
+	// +ignore=["**", "!dagger-codegen-maven-plugin/", "!dagger-java-annotation-processor/", "!dagger-java-sdk/", "!dagger-java-samples", "!LICENSE", "!README.md", "!pom.xml"]
+	sdkSourceDir *dagger.Directory,
+) (*JavaSdk, error) {
+	if sdkSourceDir == nil {
+		return nil, fmt.Errorf("sdk source directory not provided")
+	}
+	return &JavaSdk{
+		RequiredPaths: []string{},
+		SDKSourceDir:  sdkSourceDir,
+	}, nil
+}
+
+func (m *JavaSdk) Codegen(
+	ctx context.Context,
+	modSource *dagger.ModuleSource,
+	introspectionJSON *dagger.File,
+) (*dagger.GeneratedCode, error) {
+	if err := m.setModuleConfig(ctx, modSource); err != nil {
+		return nil, err
+	}
+	mvnCtr := m.codegenBase(ctx, modSource, introspectionJSON)
+
+	return dag.
+		GeneratedCode(dag.Directory().WithDirectory("/", m.generateCode(ctx, mvnCtr, introspectionJSON))).
+		WithVCSGeneratedPaths([]string{
+			"target/generated-sources/dagger/**",
+			"target/generated-sources/entrypoint/**",
+			"target/generated-sources/sdk/**",
+		}).
```

Comment:
Maybe this?
```suggestion
		WithVCSGeneratedPaths([]string{
			"target/generated-sources/**",
		}).
```

Reply:
I think I like the explicit way of describing it, but both are working.

Reply:
It's up to you, but I don't see when you'd want to have something in `generated-sources` that's not marked as "generated" 🙂 So I don't see the value in the explicitness here, and it makes it extra work to change or rename those dirs in the future, for older modules.

To be fair, even if at one point we had to add code manually to clean up these files (`.gitattributes` and `.gitignore`), the point of this API was to centralize that logic so it's probably a non issue today.